### PR TITLE
Issue #267: local bindings not visible if used in a non-hash join.

### PR DESCRIPTION
**Pull request by rbrush**
Fix for #267, ensuring a bound variable later used in a join binding is properly visible there. 

A side note: my editor cleaned up some extraneous whitespace when making this change, which normally I'd strip out and do in another commit. But I don't think it's terribly distracting and we should clean it up anyway, so I've left the whitespace cleanup in this commit as well.

### CODE REVIEW

**Comment by rbrush** [comment]
Minor update: This same problem could occur with accumulators as well, so I added a test for that too.

File: src/test/clojure/clara/test_rules.clj [hunk 1]
```diff
@@ -5452,5 +5452,25 @@
                     (insert (->Cold -10))
                     (fire-rules {:cancelling true})
                     (query cold-query))))))
-    
-  
+
+;; Test for https://github.com/cerner/clara-rules/issues/267
+(deftest local-scope-visible-in-join-filter
```
Comment (WilliamParker) [inline src/test/clojure/clara/test_rules.clj]: Nitpick: Almost all of the other tests in the file have names prefixed by "test" so we should probably make this "test-local-scope-visible-in-join-filter".

**Comment by WilliamParker** [comment]
@rbrush I've looked over this; I had some suggestions for improvements but I basically agree with the changes.

On the whitespace issue, IMO ideally we'd have some sort of automated editor-agnostic way of taking care of such issues.  I haven't tracked down where this whitespace was inserted, but I've seen different editors format Clojure differently.

**Comment by mrrodriguez** [comment]
I think Will's comments cover anything I'd add.  I did look over it though and it looks good to me.

File: src/main/clojure/clara/rules/engine.cljc [hunk 2]
```diff
@@ -646,8 +646,8 @@
       (into [type] constraints))))
 
 (defn- join-node-matches
-  [node join-filter-fn token fact env]
-  (let [beta-bindings (try (join-filter-fn token fact {})
+  [node join-filter-fn token fact fact-bindings env]
+  (let [beta-bindings (try (join-filter-fn token fact fact-bindings {})
                            (catch #?(:clj Exception :cljs :default) e
                                (throw-condition-exception {:cause e
```
Comment (rbrush) [inline src/main/clojure/clara/rules/engine.cljc]: Makes sense. I'm more tempted to go with token-bindings and fact-bindings here so we keep the distinction. 

Technically this could break a user explicitly catching an exception and checking the :bindings field, but I think that's low enough risk that it's worth using more clear names.

File: src/main/clojure/clara/rules/compiler.clj [hunk 3]
```diff
@@ -424,18 +422,29 @@
                                              [name (list accessor '?__fact__)])
                                            accessors)))
 
-        token-assignments (mapcat #(list (symbol (name %)) (list 'get-in '?__token__ [:bindings %])) binding-keys)
+        ;; Binding keys that are not newly created are drawn from the token.
+        token-binding-keys (remove element-bindings (variables-as-keywords constraints))
```
Comment (rbrush) [inline src/main/clojure/clara/rules/compiler.clj]: Yeah, we could explicitly compute this and add it as another field to the beta graph. If doing so makes things easier to debug (i.e., the small additional complexity added to the beta graph is justified), then it's worth pursuing, but I agree that is out of scope here.

Comment (rbrush) [inline src/main/clojure/clara/rules/compiler.clj]: In either case, I'll expand the comments here a bit to clarify what it's doing.

File: src/test/clojure/clara/test_rules.clj [hunk 4]
```diff
@@ -5452,5 +5452,25 @@
                     (insert (->Cold -10))
                     (fire-rules {:cancelling true})
                     (query cold-query))))))
-    
-  
+
```
Comment (rbrush) [inline src/test/clojure/clara/test_rules.clj]: I'm adding a test with the following structure, which I think is what you're looking for.

```clj
        check-reuse-previous-binding (dsl/parse-query [] [[WindSpeed (= ?t windspeed) (= ?w windspeed)]
                                                          [Temperature (= ?t temperature)
                                                           (join-filter-equals ?w ?t 10)]])
```

File: src/main/clojure/clara/rules/engine.cljc (hunk 2 above)
Comment (rbrush) [inline src/main/clojure/clara/rules/engine.cljc]: After taking this for a spin (and looking at the unit tests that check for :bindings that broke), I decided to reverse direction and just merge into the existing :bindings field. That preserves the existing behavior in most cases, keeps the structure a bit simpler, and users can easily see where the bindings came from by looking at the rule itself.

**Comment by rbrush** [comment]
The PR is updated with suggestions incorporated, and a couple new tests added as well.

**Comment by WilliamParker** [comment]
A test like this would be helpful IMO:

````
;;; query in let binding
        check-accum-result-previous-binding (dsl/parse-query []
                                                             [[?t <- (acc/min :temperature) :from [Temperature]]
                                                              [ColdAndWindy (= ?t temperature) (join-filter-equals ?t windspeed)]])
;;; assertions

(is (= [{:?t -10}]
           (-> (mk-session [check-accum-result-previous-binding])
               (insert (->Temperature -10 "MCI"))
               (insert (->ColdAndWindy -10 -10))
               fire-rules
               (query check-accum-result-previous-binding))))

    (is (empty? (-> (mk-session [check-accum-result-previous-binding])
                    (insert (->Temperature -10 "MCI"))
                    (insert (->ColdAndWindy -20 -20))
                    fire-rules
                    (query check-accum-result-previous-binding)))
        "Validate that the ?t binding from the previous accumulator result is used, rather
         than the binding in the ColdAndWindy condition that would create a ?t binding if one were
         not already present")
````
Basically the idea is to validate that a result binding is considered a "previous binding" still even when the later condition could create the binding itself.  I was specifically thinking about a binding from a result rather than a constraint in a previous condition since the paths for result bindings through the compiler are slightly different than those for variables from constraints; see [here](https://github.com/cerner/clara-rules/blob/0.13.0/src/main/clojure/clara/rules/compiler.clj#L1031) which is used to determine what bindings come from ancestors [here](https://github.com/cerner/clara-rules/blob/0.13.0/src/main/clojure/clara/rules/compiler.clj#L822).  This worked in my REPL against the branch; https://gist.github.com/WilliamParker/7e8fae758582fb52a63084f4f404f280 is a full Gist of the deftest body I used including existing cases.

Otherwise this looks good to me; +1 to merging.

**Comment by rbrush** [comment]
Alright, I went ahead and added Will's suggested test as he posted it. :)

**Comment by rbrush** [comment]
Unless there are objections, I'll merge this today since I've gotten +1's, modulo the last test I added.

**Comment by WilliamParker** [comment]
+1 to merging this @rbrush

**Comment by rbrush** [comment]
Merged! Thanks!
